### ISSUE TITLE
DbFit API redesign

### ISSUE THREAD

**Issue by javornikolov** [dbfit/dbfit#483]
This issue is to keep track of ideas which initially started emerging around the work of #463

 - [ ] `DBEnvironment` statement creation redesign #463
 - [ ] Improve packages design (e.g. currently there are objects with different purpose mixed in `util`). Consider pushing environment implementations in their own sub-packages (right now it's a common package and in theory a conflict may emerge, e.g. for some of the helper classes)
 - [ ] `EnvironmentFactory` - register with environment's class when scanning the annotations (instead of later relying on hard-coded class-name generation). That would permit moving environment classes to arbitrary package.
 - [ ] Extract interface `DbParameterAccessors` (to avoid coupling with the concrete implementation too much)
 - [ ] Rename `DBEnvironment` to `DbEnvironment` - as it's presently not consistent with the general conventions
 - [ ] Consider adding general plain-text statement creation method: in order to avoid abusing `createDdl*` for purposes other than DDL.

**Comment by MMatten** [comment]
Related to moving the `buildUpdate` command - should we move the `DELETE` statement construction out of the `Clean` fixture?

**Comment by javornikolov** [comment]
> Related to moving the `buildUpdate` command - should we move the `DELETE` statement construction out of the `Clean` fixture?

I believe yes. Added to the list.

**Comment by javornikolov** [comment]
> move the DELETE statement construction out of the Clean fixture

Pushed to #463.

**Comment by MMatten** [comment]
Regarding `DbParameterAccessor` naming:

> I'm not sure we can cay that a parameter is kind of `Attribute`.

I was trying to think of a generalised name for Stored Procedure Parameter or Table/View Column, but perhaps `Attribute` conflicts with well used Java naming. I guess if these handle `PreparedStatement` / `CallableStatement` parameters then perhaps the name of this class should be aligned with our wrapper classes for these JDBC types.

**Comment by MMatten** [comment]
Is there anything that's in the `api` package that doesn't need to be there right now?

E.g. the `Diff` interface. 

Does this need to be here at the moment? There are no references to it within the other objects in the `api` package.

**Comment by javornikolov** [comment]
In `CompareStoredQueries' we have
``` java
DataTableDiff diff = new DataTableDiff(loadRowStructure(lastRow));
```
which perhaps can be replaced with
``` java
Diff diff = new DataTableDiff(loadRowStructure(lastRow));
```
Nevertheless, perhaps it's better to move `Diff` to `diff` package.

`TestHost` is also under question - should we keep it together with  the database access api.

**Comment by MMatten** [comment]
> Nevertheless, perhaps it's better to move Diff to diff package.

Where should we put all of the types that support fixtures, but aren't part of the database access API? Should we have something like `fixtures.util`?

Or perhaps we should expand the `api` package into something like `api.dbaccess` and others for the test system interface.

**Comment by MMatten** [comment]
> Should we have something like fixtures.util?

Hmm. I guess we should only put FIT fixtures into `dbfit.fixture` as other test system implementations (SLIM, etc) will need things like `Diff`.

**Comment by javornikolov** [comment]
> Hmm. I guess we should only put FIT fixtures into `dbfit.fixture` as other test system implementations (SLIM, etc) will need things like `Diff`.

Yes. Perhaps `dbfit.fixture.report.ReportingSystem` should be moved out of `fixture` too.

Some of the points at http://theamiableapi.com/2012/01/16/java-api-design-checklist could be helpful about the package organization (the 1st 2 sections in particular - General and Naming).




> Where should we put all of the types that support fixtures, but aren't part of the database access API? Should we have something like `fixtures.util`?

Could be, but currently `Diff` and `diff` package in particular are generic (not Fit/Fitnesse-specific) so it's good to be outside `fixture`.

**Comment by javornikolov** [comment]
And seems we have a fixture in another package: `dbfit.util.ExportFixture` - which is part of the public dbfit tests syntax so perhaps it's not a good idea to move it...

**Comment by MMatten** [comment]
Also, should we look to make all fixtures interact only with objects in our client/consumer `api` package? That might simplify the interface between consumers and the database access library.

**Comment by javornikolov** [comment]
> Also, should we look to make all fixtures interact only with objects in our client/consumer api package? That might simplify the interface between consumers and the database access library.

I've been thinking about maybe avoiding direct reference to DBEnvironment methods (similar to what we have with `DbStoredProcedure`, `DeleteData`, `DbTableInsert` - some of these emerged in #463).

**Comment by MMatten** [comment]
> I've been thinking about maybe avoiding direct reference to `DBEnvironment` methods (similar to what we have with `DbStoredProcedure`, `DeleteData`, `DbTableInsert `- some of these emerged in #463).

I've been thinking along the same lines. It feels like there's a layer of objects, that sits between `DBEnvironment` and the Fit fixtures, that could emerge based upon the DbFit features. May be these would be our common API for other clients and `DBEnvironment` could be moved to an `spi` package.

**Comment by MMatten** [comment]
@javornikolov - do you still fancy moving this on so we can release v4 at some point? I have a little more time to contribute now.

**Comment by javornikolov** [comment]
@MMatten, I guess we can focus on doing the minimal stuff needed to release next version. And we can continue refactoring after that.

**Comment by MMatten** [comment]
@javornikolov, cool. I think we've already made a few API related changes since `3.2.0`. I suppose following semantic versioning, strictly, we should push out `4.0.0` but I think people might be expecting some significant new features if they see this version number. 

Do you think we can push out what we have as `3.3.0` (highlighting Informix support perhaps)? I'd like to do some work on #502 at some point soon.

**Comment by javornikolov** [comment]
Also now we have #499 which makes us incompatible with older FitNesse versions - perhaps it would be OK to go on with `4.00` directly. (Informix support sounds like a major feature anyway).

**Comment by MMatten** [comment]
Those are good points. Id be more than happy with a `4.0` release now. Can I help in preparing it?

**Comment by javornikolov** [comment]
@MMatten, it would be great if you manage to get the test VM running and check if things are working on the various databases. Also maybe #490 would be simple enough to fix it quickly and get it in the release.
Thank you!
